### PR TITLE
Jh/disable module

### PR DESCRIPTION
**Pull request by jhurst502**

### CODE REVIEW

**Comment by Cliftonz** [comment]
This pr should go to dev instead of master.
Also, it only looks like you changed one thing so I believe this should be a draft pr until the whole feature is done.

**Comment by jasekiw** [comment]
I agree, this PR is missing the migration for this property.

We need to deny requests to the UserModuleController to prevent people from being able to create a new instance of a module even if the UI denies them.

**Review by jasekiw — DISMISSED** [review]
We need to deny requests to the UserModuleController to prevent people from being able to create a new instance of a module even if the UI denies them since the UI can always be bypassed by a hacker.

File: CSLabs.Api/Controllers/UserLabController.cs [hunk 1]
```diff
@@ -72,17 +72,31 @@ public async Task<IActionResult> ProcessEndDateTime()
...
+                        .IncludeRelations()
+                        .IncludeLabHypervisor()
+                        .Include(ul => ul.Lab)
+                        .FirstAsync(ul => ul.Id == id);
+                    if (userLab.UserLabVms.Count > 0)
+                        // update the UI.
+                        return Ok(userLab);
+                    await _instantiationService.Instantiate(userLab, ProxmoxManager);
+                    await DatabaseContext.SaveChangesAsync();
+                    return Ok(userLab.GetResponse(Mapper));
```
Comment (Cliftonz) [inline CSLabs.Api/Controllers/UserLabController.cs]: You should only have one return in a function.

**Review by Cliftonz — DISMISSED** [review]
LGTM other then the two return statements

File: CSLabs.Api/Controllers/UserLabController.cs (hunk 1 above)
Comment (jhurst502) [reply CSLabs.Api/Controllers/UserLabController.cs]: I think in this case multiple returns would be fine, only one thing is being returned depending on the data. This is how we do it in the other controllers as well, and how its done according to the docs for entity framework
image

File: CSLabs.Api/Controllers/UserLabController.cs [hunk 2]
```diff
@@ -72,17 +72,31 @@ public async Task<IActionResult> ProcessEndDateTime()
...
-            await _instantiationService.Instantiate(userLab, ProxmoxManager);
-            await DatabaseContext.SaveChangesAsync();
-            return Ok(userLab.GetResponse(Mapper));
+            var query = from lab in DatabaseContext.UserLabs
+                    join usrModule in DatabaseContext.UserModules on
+                        lab.UserModuleId equals usrModule.Id
+                    join module in DatabaseContext.Modules on 
+                        usrModule.ModuleId equals module.Id
+                    where usrModule.Id == id
+                    select module.Disabled;
```
Comment (jasekiw) [inline CSLabs.Api/Controllers/UserLabController.cs]: Please use lamda version of LINQ to match the rest of the queries in the codebase.

File: CSLabs.Api/Controllers/UserLabController.cs (hunk 1 above)
Comment (jasekiw) [reply CSLabs.Api/Controllers/UserLabController.cs]: Multiple returns is fine

File: CSLabs.Api/Controllers/UserLabController.cs (hunk 2 above)
Comment (jasekiw) [reply CSLabs.Api/Controllers/UserLabController.cs]: Try the following:

```suggestion
            var isDisabled = await DatabaseContext.Modules
                .Where(m => m.Labs.Any(l => l.UserLabs.Any(ul => ul.Id == id)))
                .Select(m => m.Disabled)
                .FirstAsync();
```

File: CSLabs.Api/Controllers/UserLabController.cs [hunk 3]
```diff
@@ -72,17 +72,31 @@ public async Task<IActionResult> ProcessEndDateTime()
...
+                        .Include(ul => ul.Lab)
+                        .FirstAsync(ul => ul.Id == id);
+                    if (userLab.UserLabVms.Count > 0)
+                        // update the UI.
+                        return Ok(userLab);
+                    await _instantiationService.Instantiate(userLab, ProxmoxManager);
+                    await DatabaseContext.SaveChangesAsync();
+                    return Ok(userLab.GetResponse(Mapper));
+                }
+                return BadRequest(new {Message = "Lab is currently disabled"});
```
Comment (jasekiw) [inline CSLabs.Api/Controllers/UserLabController.cs]: I would invert this if statement to reduce nesting.

```suggestion
            if (isDisabled)
            {
                return BadRequest(new {Message = "Lab is currently disabled"});
            }
            var userLab = await DatabaseContext.UserLabs
                .IncludeRelations()
                .IncludeLabHypervisor()
                .Include(ul => ul.Lab)
                .FirstAsync(ul => ul.Id == id);
            if (userLab.UserLabVms.Count > 0) // update the UI.
                return Ok(userLab);
            await _instantiationService.Instantiate(userLab, ProxmoxManager);
            await DatabaseContext.SaveChangesAsync();
            return Ok(userLab.GetResponse(Mapper));
```

**Review by jasekiw — CHANGES_REQUESTED** [review]

**Comment by jasekiw** [comment]
@jhurst502 You did not follow the instructions properly for handling your migration, you now have two migrations, one that does something, and one that does not.

**Comment by jasekiw** [comment]
@jhurst502 You forgot the remove migration step one the second merge.
image

**Pull request merged**